src/extensions/default/SVGCodeHints/main.js

```diff
@@ -0,0 +1,273 @@
+ * copy of this software and associated documentation files (the "Software"), 
+ * to deal in the Software without restriction, including without limitation 
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, 
+ * and/or sell copies of the Software, and to permit persons to whom the 
+ * Software is furnished to do so, subject to the following conditions:
+ *  
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *  
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+ * DEALINGS IN THE SOFTWARE.
+ * 
+ */
+
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4, maxerr: 50 */
+/*global define, brackets */
+
+define(function (require, exports, module) {
+    "use strict";
+    
+    // Load dependencies.
+    var AppInit             = brackets.getModule("utils/AppInit"),
+        CodeHintManager     = brackets.getModule("editor/CodeHintManager"),
+        PreferencesManager  = brackets.getModule("preferences/PreferencesManager"),
+        XMLUtils            = brackets.getModule("language/XMLUtils"),
+        SVGTags             = require("text!SVGTags.json"),
+        SVGAttributes       = require("text!SVGAttributes.json"),
+        cachedAttributes    = {},
+        tagData,
+        attributeData,
+        isSVGEnabled;
+
+    // Define our own pref for hinting.
+    PreferencesManager.definePreference("codehint.SVGHints", "boolean", true);
+    
+    // Preferences to control hint.
+    function _isSVGHintsEnabled() {
+        return (PreferencesManager.get("codehint.SVGHints") !== false &&
+                PreferencesManager.get("showCodeHints") !== false);
+    }
+    
+    /**
+     * Returns a list of attributes used by a tag.
+     * 
+     * @param {string} tagName name of the SVG tag.
+     * @return {Array.<string>} list of attributes.
+     */
+    function getTagAttributes(tagName) {
+        var tag;
+        
+        if (!cachedAttributes.hasOwnProperty(tagName)) {
+            tag = tagData.tags[tagName];
+            cachedAttributes[tagName] = [];
+            if (tag.attributes) {
+                cachedAttributes[tagName] =  cachedAttributes[tagName].concat(tag.attributes);
+            }
+            tag.attributeGroups.forEach(function (group) {
+                if (tagData.attributeGroups.hasOwnProperty(group)) {
+                    cachedAttributes[tagName] = cachedAttributes[tagName].concat(tagData.attributeGroups[group]);
+                }
+            });
+        }
```

Comment:
This caching looks good. The only change I would make would be to sort the list here so it doesn't have to be sorted every time the cached list is used. It's ok if `getHints()` calls `sort()` on list every time since it is fast if list is already sorted.